### PR TITLE
fix(java): use `go-mvn-version` to remove `Package` duplicates

### LINKED ISSUE #7086: CycloneDX SBOM files generated by trivy contains duplicated entries with different version for the same jersey artifact
### Discussed in https://github.com/aquasecurity/trivy/discussions/7013

<div type='discussions-op-text'>

<sup>Originally posted by **schmiel** June 25, 2024</sup>
### Description

Running the `trivy` against `pulsar` image produces following output
```
    ...
    {
      "bom-ref": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41",
      "type": "library",
      "group": "org.glassfish.jersey.containers",
      "name": "jersey-container-servlet-core",
      "version": "2.41",
      "purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41",
      "properties": [
        {
          "name": "aquasecurity:trivy:FilePath",
          "value": "pulsar/lib/org.glassfish.jersey.containers-jersey-container-servlet-core-2.41.jar"
        },
        {
          "name": "aquasecurity:trivy:LayerDiffID",
          "value": "sha256:f56c34161ed13ae336a78c5ac6902265890265db8431c0534aa2715f4d679129"
        },
        {
          "name": "aquasecurity:trivy:LayerDigest",
          "value": "sha256:de88e9845ffb8d679749bcc839a302b86b2ff2f6394f9731653cc5b21efa6bd7"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "jar"
        }
      ]
    },
    {
      "bom-ref": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41.0",
      "type": "library",
      "group": "org.glassfish.jersey.containers",
      "name": "jersey-container-servlet-core",
      "version": "2.41.0",
      "purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41.0",
      "properties": [
        {
          "name": "aquasecurity:trivy:FilePath",
          "value": "pulsar/lib/org.glassfish.jersey.containers-jersey-container-servlet-core-2.41.jar"
        },
        {
          "name": "aquasecurity:trivy:LayerDiffID",
          "value": "sha256:f56c34161ed13ae336a78c5ac6902265890265db8431c0534aa2715f4d679129"
        },
        {
          "name": "aquasecurity:trivy:LayerDigest",
          "value": "sha256:de88e9845ffb8d679749bcc839a302b86b2ff2f6394f9731653cc5b21efa6bd7"
        },
        {
          "name": "aquasecurity:trivy:PkgType",
          "value": "jar"
        }
      ]
    },
    ...
```
For the same file with exactly the same sha two entries are generated.

### Desired Behavior

Only the entry with version `2.41` is generated.

### Actual Behavior

Two entries for the same artifact are generated, correct one with version `2.41` and the wrong one with version `2.41.0`

### Reproduction Steps

```bash
1.Run trivy against pulsar docker image: `trivy image apachepulsar/pulsar:3.3.0 --format cyclonedx`
2. Check the result - duplicated entries are visible for a few jersey artifacts, namely: jersey-container-servlet-core, jersey-container-servlet, jersey-media-json-jackson and jersey-media-multipart
```

### Target

Container Image

### Scanner

None

### Output Format

CycloneDX

### Mode

Standalone

### Debug Output

```bash
trivy image apachepulsar/pulsar:3.3.0 --format cyclonedx --debug
```

### Operating System

Ubuntu 20.04

### Version

```bash
./trivy version                                                   
Version: 0.52.2
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-06-25 12:11:52.331411989 +0000 UTC
  NextUpdate: 2024-06-25 18:11:52.331411718 +0000 UTC
  DownloadedAt: 2024-06-25 14:13:57.518535772 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2024-06-25 01:09:07.930189823 +0000 UTC
  NextUpdate: 2024-06-28 01:09:07.930189662 +0000 UTC
  DownloadedAt: 2024-06-25 09:18:40.719903085 +0000 UTC
```

### Checklist

- [X] Run `trivy clean --all`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)</div>

### PR DESCRIPTION
**Pull request by DmitriyLewen**
## Description
We [remove duplicates of packages](https://github.com/aquasecurity/trivy/blob/266d9b1f4bd2e89f1d58a16698f755a1734c644e/pkg/dependency/parser/java/jar/parse.go#L441-L445).
But there are cases when Packages uses same version, but one of package omits `0` patch version (e.g. `2.17.0` and `2.17`).

Using `go-mvn-version` to compare version solves this problem.

before:
```bash
➜ trivy -q image apachepulsar/pulsar:3.3.0 --format cyclonedx | grep '"purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core'
      "purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41",
      "purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41.0",
```

after:
```bash
➜ trivy -q image apachepulsar/pulsar:3.3.0 --format cyclonedx | grep '"purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core'
      "purl": "pkg:maven/org.glassfish.jersey.containers/jersey-container-servlet-core@2.41",
```

## Related issues
- Close #7086

## Checklist
- [x] I've read the [guidelines for contributing](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/) to this repository.
- [x] I've followed the [conventions](https://aquasecurity.github.io/trivy/latest/community/contribute/pr/#title) in the PR title.
- [x] I've added tests that prove my fix is effective or that my feature works.
- [ ] I've updated the [documentation](https://github.com/aquasecurity/trivy/blob/main/docs) with the relevant information (if needed).
- [ ] I've added usage information (if the PR introduces new options)
- [x] I've included a "before" and "after" example to the description (if the PR is a user interface change).

### CODE REVIEW

File: pkg/dependency/parser/java/jar/parse.go [hunk 1]
```diff
@@ -439,7 +439,24 @@ func (m manifest) determineVersion() (string, error) {
...
+		// cf. https://github.com/masahiro331/go-mvn-version/blob/d3157d602a08806ad94464c443e0cef1370694a1/version.go#L20-L25
+		pkgVer, _ := mavenversion.NewVersion(pkg.Version)
+		savedVers, ok := uniq[uniqID]
+		if !ok || !slices.ContainsFunc(savedVers, func(v mavenversion.Version) bool {
+			// There are times when patch `0` is omitted.
+			// So we can't compare versions just as strings
+			// for example `2.17.0` and `2.17` must be equal
+			return v.Equal(pkgVer)
+		}) {
+			uniq[uniqID] = []mavenversion.Version{pkgVer}
```
Comment (knqyf263) [inline pkg/dependency/parser/java/jar/parse.go:457]: Don't we have to append the new version here?
```suggestion
			uniq[uniqID] = append(uniq[uniqID], pkgVer)
```

Comment (DmitriyLewen) [reply pkg/dependency/parser/java/jar/parse.go]: right! thanks!

Comment (knqyf263) [reply pkg/dependency/parser/java/jar/parse.go]: We should add a test so we will not have regression.

Comment (DmitriyLewen) [reply pkg/dependency/parser/java/jar/parse.go]: Added `pom.properties` files for that in `io.quarkus.gizmo.gizmo-1.1.jar` file.
New structure of `io.quarkus.gizmo.gizmo-1.1.jar`:
https://github.com/aquasecurity/trivy/blob/7a5018e1d943839b85ea9a7df6404f82f1aaad89/pkg/dependency/parser/java/jar/parse_test.go#L172-L193

**Review by knqyf263 — APPROVED** [review]

**Pull request merged**